src/main/scala/net/snowflake/spark/snowflake/io/StageWriter.scala

```diff
@@ -468,66 +482,115 @@ private[io] object StageWriter {
       firstCopyFileSet.get.toSet
     )
     log.debug(Utils.sanitizeQueryText(copyStatement.toString))
+    var lastStatement = copyStatement.toString
+    logAndAppend(progress, s"First COPY command: $lastStatement")
 
-    // execute the COPY INTO TABLE statement
-    val resultSet = copyStatement.execute(params.bindVariableEnabled)(conn)
-
-    // Save the original COPY command even if additional COPY is run.
-    Utils.setLastCopyLoad(copyStatement.toString)
-
-    // Get missed files if there are any.
-    var missedFileSet = getCopyMissedFiles(params, resultSet, expectedFileSet)
-
-    // If any files are missed, execute 2nd COPY command
-    if (missedFileSet.nonEmpty) {
-      // Negative test:
-      // Only load part of missed files.
-      // Exception is raised for the failure.
-      val secondCopyFileSet: Option[mutable.Set[String]] =
-      if (TestHook.isTestFlagEnabled(
-        TestHookFlag.TH_COPY_INTO_TABLE_MISS_FILES_FAIL)) {
-        Some(missedFileSet.grouped(2).toList.head)
-      } else {
-        Some(missedFileSet)
-      }
-
-      // Generate copy command with missed files only
-      useFilesClause = true
-      val copyWithFileClause = StageWriter.copySql(
-        schema,
-        saveMode,
-        params,
-        targetTable,
-        file,
-        tempStage,
-        format,
-        conn,
-        useFilesClause,
-        secondCopyFileSet.get.toSet
-      )
-
-      def getMissedFileInfo(missedFileSet: mutable.Set[String]): String = {
-        s"""missedFileCount=${missedFileSet.size}
-           | Files: (${missedFileSet.mkString(", ")})
-           |""".stripMargin.filter(_ >= ' ')
+    var lastQueryID: String = null
+    try {
+      // execute the COPY INTO TABLE statement
+      val (resultSet, queryID, errorThrow) =
```

Comment:
This is super-ugly, you shouldn't have to change this stuff,

Reply:
It's some complex, but I think it is very useful for diagnosis.
I can just change to call executeWithoutThrow(), check error and send telemetry message.
That will be easier, but I think the current is better.
With current change, if COPY fails, we can get details info about the progress and situation.